File: src/main/kotlin/xyz/mahmoudahmed/dsl/NestedTargetPathBuilder.java
package xyz.mahmoudahmed.dsl;

import java.util.function.Function;

public class NestedTargetPathBuilder {
    private final NestedMappingBuilder parent;
    private final String sourcePath;
    private final String targetPath;
    private final Function<Object, Object> transformation;

    NestedTargetPathBuilder(NestedMappingBuilder parent, String sourcePath, String targetPath,
                            Function<Object, Object> transformation) {
        this.parent = parent;
        this.sourcePath = sourcePath;
        this.targetPath = targetPath;
        this.transformation = transformation;
    }

    public NestedMappingBuilder end() {
        MappingRule rule = new NestedSimpleMapping(
                sourcePath,
                parent.getTargetCollection() + (targetPath.startsWith(".") ? "" : ".") + targetPath,
                transformation,
                null
        );
        parent.addNestedRule(rule);
        return parent;
    }
}

File: src/main/kotlin/xyz/mahmoudahmed/dsl/collections/NestedTargetPathBuilder.kt
import xyz.mahmoudahmed.dsl.collections.NestedMappingBuilder
import xyz.mahmoudahmed.dsl.collections.NestedSimpleMapping

class NestedTargetPathBuilder(
    private val parent: NestedMappingBuilder,
    private val sourcePath: String,
    private val targetPath: String,
    private val transformation: ((Any) -> Any)?
) {
    /**
     * Finishes the mapping configuration.
     *
     * @return The parent nested mapping builder
     */
    fun end(): NestedMappingBuilder {
        // Just use the targetPath directly, no need to prefix it
        val rule = NestedSimpleMapping<Any>(
            sourcePath,
            targetPath,  // Use targetPath directly
            transformation,
            null
        )
        parent.addNestedRule(rule)
        return parent
    }

    private fun buildTargetPath(): String {
        return targetPath
    }
}